backend/file-ucloud-service/src/jvmMain/kotlin/services/FileScanner.kt

```diff
@@ -0,0 +1,505 @@
+                            null
+                        }
+                    }
+
+                } else chunk
+
+                while (true) {
+                    collectionsList.zip(filteredChunk).forEach { (coll, path) ->
+                        withContext(pool) {
+                            launch {
+                                val file =
+                                    pathConverter.relativeToInternal(RelativeInternalFile("/projects/$path}"))
+                                submitScan(file, coll)
+                            }.join()
+                        }
+                    }
+
+                    if (resolvedCollections.next == null) {
+                        return@forEach
+                    }
+                    resolvedCollections = retrieveCollections(
+                        providerGenerated = true,
+                        includeOthers = true,
+                        chunk.map { PathConverter.COLLECTION_PROJECT_PREFIX + it },
+                        resolvedCollections.next
+                    )
+                }
+            }
+
+            if (session.sendPreparedStatement(
+                """
+                    UPDATE file_ucloud.storage_scan set
+                    last_system_scan = now()
+                    where true
+                """
+            ).rowsAffected == 0L ) {
+                session.sendPreparedStatement(
+                    """
+                        INSERT into file_ucloud.storage_scan
+                        values (now())
+                    """.trimIndent()
+                )
+            }
+            println("Scan done")
+        }
+    }
+
+    //Loses microseconds from timestamp
+    private fun rctimeToApproxLocalDateTime(rctime: String): LocalDateTime {
+        return LocalDateTime(Date(rctime.split(".").first().toLong()*1000))
+    }
+
+    private suspend fun submitScan(file: InternalFile, collection: FileCollection, upperLimitOfEntries: Long = Long.MAX_VALUE) {
+        val fileList = fs.listFiles(file).map { InternalFile(file.path + "/" + it) }
+        if (fileList.isEmpty() || fileList.any { it.fileName() == ".skipFolder" }) {
+            log.debug("Skipping ${file.path} due to .skipFolder file or because the folder is empty")
+            return
+        }
+        if (rctimeToApproxLocalDateTime(fastDirectoryStats.getRecursiveTime(file)) < lastScan ) {
+            log.debug("rcTime is older than lastscan = no change")
+            return
+        }
+
+        val thisFileInIndex = run {
+            val source =
+                elastic.get(GetRequest(FILES_INDEX, pathConverter.internalToUCloud(file).path), RequestOptions.DEFAULT)?.sourceAsString
+            if (source != null) {
+                defaultMapper.decodeFromString<ElasticIndexedFile>(source)
+            } else {
+                null
+            }
+        }
+
+        var newUpperLimitOfEntries = 1L
+        if (File(file.path).isDirectory) {
```

Comment:
Should be changed to use the `NativeFs` interface instead (to correctly handle symlinks)